source/robot_model/src/Model.cpp

```diff
@@ -244,9 +244,76 @@ std::vector<state_representation::CartesianPose> Model::forward_geometry(
   return this->forward_geometry(joint_state, frame_ids);
 }
 
-state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState&) const {
-  // TODO
-  return state_representation::JointPositions();
+state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState& desired_cartesian_state,
+                                                            const state_representation::JointState& current_joint_state,
+                                                            std::string frame_name, const double& tolerance, const int& max_number_of_iteration) {
+
+  if (frame_name.empty()) {
+    // get last frame if none specified
+    frame_name = this->robot_model_.frames.back().name;
+  }else {
+    // throw error if specified frame does not exist
+    if (!this->robot_model_.existFrame(frame_name)) { throw (exceptions::FrameNotFoundException(frame_name)); }
```

Comment:
I think there are still some formatting issues. Are you using auto formatting with our style guide ?